### PR TITLE
Provider generics interface to handle non-png icons

### PR DESCRIPTION
**Pull request by jmatsu**
Waiting for

- [x] https://github.com/DeployGate/android_apk/pull/33
- [x] https://github.com/DeployGate/android_apk/pull/34

Breaking changes

- `backward_compatible_adaptive_icon?` now returns true whenever it's configured properly.
- Drop setters of `icon`, `adaptive_icon, `backward_compatible_adaptive_icon`

New itnerface

- AppIcon class: this provides IO interface to get app icons
- AndroidApk#app_icons: this returns an array sorted by resolution priority
- Xmltree: this is not a XML document but tree provided by aapt/aapt2

### CODE REVIEW

File: .rubocop.yml [hunk 1]
```diff
@@ -168,3 +168,6 @@ Style/TrailingCommaInArrayLiteral:
 
 Style/TrailingCommaInHashLiteral:
   Enabled: false
+
+Style/AccessModifierDeclarations:
```
Comment (jmatsu) [inline .rubocop.yml:172]: https://www.rubydoc.info/gems/rubocop/RuboCop/Cop/Style/AccessModifierDeclarations

File: lib/android_apk.rb [hunk 2]
```diff
@@ -421,28 +455,11 @@ def self.read_signature(apk, filepath)
...
-
-    # at least one png icon is required if min sdk version doesn't support adaptive icon
-    if apk.min_sdk_version.to_i < ADAPTIVE_ICON_SDK && png_dpi.nil?
-      return
-    end
-
-    # invalid xml file may throw an error
-    Zip::File.open(filepath) do |zip_file|
-      apk.adaptive_icon = zip_file.find_entry(adaptive_icon_path)&.get_input_stream do |entry|
-        !!entry.read.include?("adaptive-icon")
```
Comment (jmatsu) [inline lib/android_apk.rb:440]: this logic worked even if the xml is decoded but the result included false-positives.

File: android_apk.gemspec [hunk 3]
```diff
@@ -16,7 +16,6 @@ Gem::Specification.new do |spec|
   spec.required_ruby_version = ">= 2.6.0"
 
   spec.files = `git ls-files | grep -v 'spec/fixture'`.split($/)
-  spec.test_files = spec.files.grep(%r{^(test|spec)/})
```
Comment (jmatsu) [inline android_apk.gemspec:19]: https://github.com/rubygems/guides/issues/90#issuecomment-49213972

File: lib/android_apk.rb [hunk 4]
```diff
@@ -178,27 +171,43 @@ def self.analyze(filepath)
...
 
   def initialize
-    self.adaptive_icon = false
-    self.backward_compatible_adaptive_icon = false
     self.verified = false
     self.test_only = false
   end
 
+  # @return [Array<AndroidApk::AppIcon>]
+  def app_icons
```
Comment (jmatsu) [inline lib/android_apk.rb:192]: We will check this array from 0 to last to look up icons.

File: lib/android_apk/xmltree.rb [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    end
+
+    def initialize(content:)
+      @root_element = nil
+
+      line_num = 1
+      content&.split(/\n/) do |line|
+        line.strip!
+
+        if line.start_with?("E: ")
```
Comment (jmatsu) [inline lib/android_apk/xmltree.rb:26]: The first `E: `-prefixed line indicates the root element of the XML.

File: lib/android_apk/xmltree.rb [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+        stdout, _, status = Open3.capture3("aapt", "dump", "xmltree", apk_filepath, xml_filepath)
+        stdout if status.success?
+      end
+    end
+
+    def initialize(content:)
+      @root_element = nil
+
+      line_num = 1
+      content&.split(/\n/) do |line|
```
Comment (jmatsu) [inline lib/android_apk/xmltree.rb:23]: tip: `split` with block requires Ruby 2.6

**Review by anoworl — APPROVED** [review]
LGTM!

**Comment by jmatsu** [comment]
Thank you so much~~~!

**Pull request merged**